[WIP] Native TOML config

ziima opened the pull request:
This is a prototype solution for #999, see discussion there.

This branch will be rebased a lot.

gaborbernat commented on src/tox/config/source/toml.py:
```diff
@@ -0,0 +1,142 @@
+        return f"{type(self).__name__}(path={self.path})"
+
+    def transform_section(self, section: Section) -> Section:  # noqa: PLR6301
+        return IniSection(section.prefix, section.name)
```
We should not use the ini system at all.

ziima replied:
I duplicated `IniSection` to `TomlSection` for the time being. I plan to do more clean up later on.

gaborbernat commented on src/tox/config/source/toml.py:
```diff
@@ -0,0 +1,142 @@
+            if in_section.prefix is not None:  # no prefix specified, so this could imply our own prefix
+                yield IniSection(in_section.prefix, a_base)
+
+    def sections(self) -> Iterator[IniSection]:
```
It makes no sense why have anything to do with the ini system here.

gaborbernat commented on src/tox/config/source/toml.py:
```diff
@@ -0,0 +1,142 @@
+    This is experimental API! Expect things to be broken.
+    """
+
+    CORE_SECTION = CORE
```
I don't see why we should reuse the ini systems prefixes for both the core and testing environment.

gaborbernat commented on tests/config/source/test_discover.py:
```diff
@@ -10,7 +10,7 @@
 def out_no_src(path: Path) -> str:
     return (
-        f"ROOT: No tox.ini or setup.cfg or pyproject.toml found, assuming empty tox.ini at {path}\n"
+        f"ROOT: No tox.ini or tox.toml or setup.cfg or pyproject.toml found, assuming empty tox.ini at {path}\n"
```
I'm not sure I agree with the need for a dedicated TOML file.  I am currently -1 on adding that.

ssbarnea replied:
I am in favour of defaulting to `pyproject.toml` but still allowing user to define an alternative config file to load. Some might find a need for it (like CI/CD, or custom user modified one when they cannot merge their changes in the default one).

I would say that even with alternative filename, I would still expect it to follow the same sections as the `pyproject.toml`, just to keep code simple.

ziima replied:
I went through the docs for several non-build tools I know (excluded those which do not support `pyproject.toml` at all) to see if the support custom TOML file. Here are results:

| Tool | Custom TOML file | Custom TOML root |
| ---- | --------------------- | ------------------ |
| mypy | none | |
| ruff | `ruff.toml` | none |
| bump-my-version | `.bumpversion.toml` | `tool.bumpversion` or `bumpversion` |
| isort | none | |
| black | none | |
| codespell | none | |
| coverage | none | |
| towncrier | `towncrier.toml` | `tool.towncrier` |

I'd say custom TOML file is not that rare. Custom root in TOML file seems to vary.

gaborbernat commented on tests/config/source/test_source_toml.py:
```diff
@@ -0,0 +1,83 @@
+    )
+    assert custom_section["a"] == "b"
+
+
```
I'm not a fan of this long comments.

ziima commented:
I've ditched `TomlLoader` and updated `MemoryLoader` instead. They were mostly the same.

gaborbernat commented on src/tox/config/source/toml.py:
```diff
@@ -0,0 +1,140 @@
+
+
+class TomlSource(Source):
+    """Configuration sourced from a toml file (such as tox.toml).
```
I do not want a dedicated toml, but should be inside pyproject.toml.

stinovlas replied:
I'd prefer to have an option of separate TOML config file as well. If you don't like a dedicated file such as `tox.toml`, what about allowing to specify configuration file in env variable, such as `TOX_CONFIG`?

gaborbernat replied:
I don't like the environment idea 😃

ziima commented:
I changed testenv in TOML to `tool.tox.env` as suggested in https://github.com/tox-dev/tox/issues/999#issuecomment-2231261265

Currently supported pyproject.toml looks like 
```toml
[tool.tox]
env_list = ['py311', 'py312', 'type']

# The default testenv is `testenv`.
[tool.tox.env.testenv]
deps = 'pytest'
commands = ['pytest tests']

[tool.tox.env.type]
deps = 'mypy'
commands = ['mypy src']
```

I also added support for native `pyproject.toml`. It broke ability to load legacy `pyproject.toml` using `--conf` though. To be fixed later.

I split `BaseSection` from `Section` to handle different structure of data within TOML. I'm not quite sure what to do with that at the moment.

I keep `tox.toml` support at least for tests during development. It's easier that way.

gaborbernat commented:
I think deps needs to be a list of strings. Also, I would like to rename the base environment name to something like testenv -> base.

ziima commented:
Ad options: `deps` just needs a tweak to work, but there is an underlying issue regarding factors. I'm not yet sure how to handle factors in TOML. For `deps`, I could just do
```toml
deps = ["quality: ruff"]
```
but that is somewhat clumsy and it really doesn't work with non-string options, such as `ignore_errors`.

I'm starting to think the best approach would be to provide a subkey to explicitly define factors in TOML, such as

```toml
[[tox.deps]]
name = "ruff"
when = ["quality"]

[[tox.deps]]
name = "pip"
when = ["py312", "py311"]

[[tox.deps]]
name = "setuptools"

# or a shortcut of
[tox]
deps = ["ruff", "pip", "setuptools"]
# when none of the deps has custom factors.
```

Ad base environment name: I tried to respect the `base` option, which has `testenv` as default value. I haven't polish this part, but this is the goal for the moment.

gaborbernat commented:
IMHO we should do something similar to what hatch does for environment generation and factors.

gaborbernat commented:
@ziima do you plan to finalize this?

ziima commented:
I'm back from holiday, so I should be able to find some time to work on this again.

gaborbernat commented:
Great to hear so. And thank you for doing all the hard work.

gaborbernat commented:
Replaced by https://github.com/tox-dev/tox/pull/3353
